keps/sig-scheduling/3094-pod-topology-spread-considering-taints/README.md

```diff
@@ -0,0 +1,836 @@
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+Introduce a new field to `TopologySpreadConstraint` to define all node inclusion
+policies including nodeAffinity and nodeTaint.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+When calculating pod topology spread skew, I want to exclude nodes that don't tolerate
+all taints to prevent pods from falling into unexpected Pending state.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+- Checking nodeAffinity and nodeTaints the same time may lead to performance
+problem, we need to verify this by adding performance tests. If performance problem
+does exists, we'd like to add a toleration parser and cache the parsed object
+during PreFilter.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+A new field named `NodeInclusionPolicies` will be introduced to `TopologySpreadConstraint`:
+```golang
+type TopologySpreadConstraint struct {
+  // NodeInclusionPolicies includes several policies
+  // when calculating pod topology spread skew
+  NodeInclusionPolicies NodeInclusionPolicies
+}
+```
+
+There are two policies now regarding to nodeAffinity and nodeTaint:
+```golang
+type NodeInclusionPolicies struct {
+  // NodeAffinity policy indicates how we will treat nodeAffinity/nodeSelector
+  // when calculating pod topology spread skew. The value could be Respect/Ignore.
+  // By default, the policy is Respect.
+  NodeAffinity PolicyName
+  // NodeTaint policy indicates how we will treat node taints
+  // when calculating pod topology spread skew. The value could be Respect/Ignore.
+  // By default, the policy is Ignore for backwards compatibility.
```

Comment:
// when calculating pod topology spread skew. The semantics vary by PolicyName:
  // - Respect: tainted nodes that tolerate the incoming pod, along with regular nodes, will be included.
  // - Ignore (default): all nodes will be included.